orttraining/orttraining/test/training_ops/cpu/tensor/gather_grad_op_test.cc

```diff
@@ -14,6 +14,54 @@ namespace onnxruntime {
 namespace test {
 
 namespace {
+
+std::pair<std::vector<int64_t>, std::vector<int64_t>> CalculateSortedIndices(const std::vector<int64_t>& indices)
+{
+    std::vector<int64_t> dY_indices_sorted(indices.size());
+    std::iota(dY_indices_sorted.begin(), dY_indices_sorted.end(), 0);
+    sort(dY_indices_sorted.begin(), dY_indices_sorted.end(), [&indices](const size_t& i, const size_t& j) { return indices[i] < indices[j]; } );
+    std::vector<int64_t> dX_indices_sorted(indices);
+    sort(dX_indices_sorted.begin(), dX_indices_sorted.end());
+
+    return {dX_indices_sorted, dY_indices_sorted};
+}
+
+std::pair<int32_t, std::vector<int32_t>> CalculateNumberOfSegmentsAndSegmentOffsets(const std::vector<int64_t>& sorted_indices)
+{
+    int32_t num_segments = 1;
+    std::vector<int32_t> segment_offsets({0});
+
+    for (size_t i = 1; i < sorted_indices.size(); ++i)
+    {
+        if (sorted_indices[i] != sorted_indices[i-1])
+        {
+            num_segments++;
+            segment_offsets.push_back(static_cast<int32_t>(i));
+        }
+    }
+
+    return {num_segments, segment_offsets};
```

Comment:
num_segments == segment_offsets.size() all the time, returning std::vector<int32_t> should be enough.

Reply:
You're right. The `num_segments` would always be equal to `(int32_t)segment_offsets.size()`. But I think since this is a unit test, to have `num_segments` explicitly returned might be clearer to the reader of the code.

Reply:
conceptually, we have one segment offset for each segment right?